### PR TITLE
MAINT: Flight class post methods

### PR DESCRIPTION
**Pull request by giovaniceotto**
## Pull request type

- [x] Code maintenance (refactoring, formatting, renaming, tests)

## Pull request checklist

- Code base maintenance (refactoring, formatting, renaming):

  - [x] Docs have been reviewed and added / updated if needed
  - [x] Lint (`black rocketpy`) has passed locally and any fixes were made
  - [x] All tests (`pytest --runslow`) have passed locally

## What is the current behavior?

Most methods in the Flight class which can only be calculated after the simulation is done make use of the `cached_property` decorator.

## What is the new behavior?

Whenever possible, the `cahced_property` decorator was replaced by the `rocketpy.Function.funcify_method` decorator, which significantly improves code readability, maintenance and reduces lines of code. This was last discussed on #283 with @Gui-FernandesBR.

Further minor improvements include:

- Creation of `Flight.time` property to avoid having to retrieve the time array as `self.x[:, 0]`. This is a great improvement to code readability.
- Creation of `Flight.solutionArray`, which stores an `np.array` version of the solution list. This is more efficient since the conversion from list to `np.array` only needs to be done once.
- Creation of `Flight.outOfRailTimeIndex`. Before, this was calculated many times in different methods, which is not at all efficient. Now, it is saved once the simulation detects the out of rail event. Thus, no more array searches are needed!
- Creation of `Flight._frequency_analysis` to reduce code repetition for Frequency Response methods.

**In total, a net total of 755 lines of code were removed by this PR, decreasing the size of `Flight.py` by 15%**, while keeping the exact same functionality.

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

Waiting for #291 before exiting draft status.

### CODE REVIEW

**Comment by giovaniceotto** [comment]
@Gui-FernandesBR, this is ready for review!

Don't forget to read "What is the new behavior?" section of the PR message to get a sense of what was done.

I would be happy to join you in a discord voice chat to answer any questions and explain more details of this PR. But the fact that all tests are passing (and none had to be modified) is definitely a good sign!

**Comment by giovaniceotto** [comment]
Problems identified so far:

- Dynamic Stability Analysis section of the Getting Started Notebook not working as intended. Seems to be a problem that came before this PR, as the same problem manifests itself in the master branch.
image

- Frequency response analysis is different from previous results. I am investigating which one is correct, but help is wanted.

**Comment by Gui-FernandesBR** [comment]
> Problems identified so far:
> 
> * Dynamic Stability Analysis section of the Getting Started Notebook not working as intended. Seems to be a problem that came before this PR, as the same problem manifests itself in the master branch.
>   image
> * Frequency response analysis is different from previous results. I am investigating which one is correct, but help is wanted.

the variable "factor" is not being used in the loop. Adding the following term in the code allowed me to get expected behavior:
image
image

I will try ingestigate frequency problem as well!

**Comment by giovaniceotto** [comment]
Alright, since the factor issue not being related to this PR, I think we need a new PR for it that goes directly to master.

File: rocketpy/Flight.py
Comment (Gui-FernandesBR) [inline rocketpy/Flight.py:2675]: After this row, I can no longer understand the old (latest) implementation, therefore I cannot analyze what you (possibly) did wrong in your new proposal.

**Review by Gui-FernandesBR — COMMENTED** [review]
Hey @giovaniceotto , really nice one.

I loved the final result. It is much easier to understand  the flow of the code now.

I tried to find the bug you pointed regarding the frequency analysis, but I was not able to find what is wrong. Could you share a **current vs expected** behavior comparision pls?

In case we need further investigations on that, I'd like to request the discord meeting you offer, otherwise we can bring someone else to analyze it.

File: rocketpy/Flight.py
Comment (giovaniceotto) [reply rocketpy/Flight.py:1650]: Agreed!

File: rocketpy/Flight.py
Comment (giovaniceotto) [reply rocketpy/Flight.py:2091]: Not really sure... I guess I just thought the higher the better.

**Comment by giovaniceotto** [comment]
My bad, @Gui-FernandesBR. I have checked both codes side by side and the produce the exact same output if given the same input. The only difference came from different inputs. Namely, the interval of the input signal was different.

I have set a default window which I judge is appropriate for most rocket flights. This is not perfect, but it is already an improvement to the previous implementation. As this is not the main focus of this PR, I suggest we move forward.

You can check the Getting Started Notebook and all unit tests to make sure all is in order.

File: rocketpy/Flight.py
Comment (Gui-FernandesBR) [reply rocketpy/Flight.py]: @giovaniceotto , I ask again, don't you thnik this method should receive double underscore? This way you would create a non public method. Otherwise it just feels weird the method 's name

Comment (giovaniceotto) [reply rocketpy/Flight.py]: Sorry, hadn't noticed the question. I did not want the method to be private because it can be used with any Function. Using a single underscore means it is meant to be used internally (therefore if we change it later, it is not a breaking change), but can also be used outside of the class by other functions and other classes (think analysis class).

I would prefer one of the following:
1. Keep it this way;
2. Make it public, by removing the leading underscore;
3. Transfer its functionality into the Function class (Function.toFrequencyDomain perhaps).

The third option is out of scope in relation to this PR. If you chose one of the other two, we could make the change and move forward.

Comment (Gui-FernandesBR) [reply rocketpy/Flight.py]: I'm sorry to bother you again, but I don't fell too happy with this method yet.
Seems to me a kind impossible to use it by another class, using the curret implementation, as you can only use it if you create a Flight object

**Some considerations here:**
1. Consider the recently #283 comments, the appropiate place for functions that can be used by different classes is tools.py. 
2. However, at liquid motors project, they are using supplement.py instead.
3. Yeah, confusion and confusion!

**My suggestions:** 
**Option A**: remove the depency of self.outOfRailTime attribute, wich is the only class attribute accessed by this method. We can set this as an argument and therefore the method is ready to become a `@staticmethod`, without depending on self as an argument.

**Option B**: Just move it to tools.py and invoke here at function.

Comment (giovaniceotto) [reply rocketpy/Flight.py]: Unfortunately, I was really confusing when explaining my motives. An analysis class can make use of this method to get the frequency analysis of any Flight class attribute. I did not mean that it could use to perform frequency analysis of other signals.

I will follow your suggestion and remove the dependency of the `Flight.outOfRailTime` attribute and create a new commit.

**Quick note**:

> However, at liquid motors project, they are using supplement.py instead.

As of #299, this is not the case anymore.

**Comment by giovaniceotto** [comment]
Here is Copilot's suggestion:
image

**Comment by Gui-FernandesBR** [comment]
> Here is Copilot's suggestion: image

If copilot says, I have no reason to disagree!

**Comment by giovaniceotto** [comment]
@Gui-FernandesBR, tests are not passing through GitHub Actions still, but locally everything seems fine...

**Comment by Gui-FernandesBR** [comment]
I got this result when running the getting started notebook with default settings:

image

Is this the correct expected behavior?

**Review by Gui-FernandesBR — APPROVED** [review]
Yeah Sr., this one seems to be one of the best additions to the next version. Code is cleaner and faster, goodbye postProcess method!

I confirm that tests are running locally in my machine (windows). Getting started noteook is also running correctly, **with the exception of frequencyResponses plots**. Let's move forward then. I'll let you decide if we proceed right now or invest more time understanding frequencyResponses again, but imo the overall goals were done by your PR, it's just a matter of refining the solution now.

**Comment by MateusStano** [comment]
> I got this result when running the getting started notebook with default settings:
> 
> image
> 
> Is this the correct expected behavior?

This is definitely wrong lol
Im looking into it

File: rocketpy/Function.py
Comment (MateusStano) [inline rocketpy/Function.py:1128]:
```suggestion
        FourierAmplitude = np.abs(np.fft.fft(sampledPoints) / (numberOfSamples) / 2)
```

Comment (MateusStano) [reply rocketpy/Function.py:1128]: Small error here. The magnitude from `np.fft.fft` comes multiplied by the `numberOfSamples/2`

Comment (Gui-FernandesBR) [reply rocketpy/Function.py:1128]: That's definitely a good catch

File: rocketpy/Flight.py
Comment (MateusStano) [inline rocketpy/Flight.py:2094]: Defining the upper time bound to be just 5 seconds after the lower bound is what is making the weird graph shown in `Flight.allInfo()` that @Gui-FernandesBR  showed. This makes senses makes senses because smaller time frames will make the fft more 'broad' (or whatever the correct term is)

So the graph is right, but I don't know if having this is very useful compared to what we used before.

Just to exemplify:

**omega1FrequencyResponse in a 5 sec window:**

Figure 12

**With xlim to 5hz:**

Figure 14

**omega1FrequencyResponse until Flight.tFinal (how it was before):**

Figure 13

**With xlim to 5hz:**

Figure 15


I think it makes more sense to look at the frequency response of the entire flight in this case. Or maybe just make the window bigger than 5 seconds

What do you guys think @giovaniceotto @Gui-FernandesBR ?

**Comment by MateusStano** [comment]
Also, should this be going to develop? It is a lot of changes...

**Comment by Gui-FernandesBR** [comment]
> Also, should this be going to develop? It is a lot of changes...

The post methods are needed, otherwise current develop version doesn't work properly... Properties have been defined with arguments ([example here](https://github.com/RocketPy-Team/RocketPy/blob/develop/rocketpy/Flight.py#L1615)), and this need to be changed before releasing anything else.

**Comment by giovaniceotto** [comment]
I agree that this should head directly to develop. Even though there are many changes, there are no breaking changes, and everything has been reviewed thoroughly. 

Regarding the frequency analysis plots, they are remarkably similar to what is currently released to be honest.

I do not agree with performing the FFT over the entire flight range, since after parachute ejection, the angular velocities are kept constant. What we could do is perform it from out of rail until the first parachute event, but even that I disagree with. The main reason is that the signal is heavily damped, and its frequency also changes widely during the flight since it is a function of velocity. Therefore, performing a FFT over a wide domain does us no good. Since we are usually interested in the behaviour right after the rail, I chose to perform the analysis until 5 seconds, which is what is done in the Getting Started Notebook for a while now.

The only way to solve this once and for always is to implement time-frequency analysis methods, such as STFT (short-time Fourier transform), wavelets decomposition, least square spectral analysis or Prony's method. These are very interesting and would be an welcome addition. However, they are certainly out of scope for this PR.

In conclusion, I vote for us to raise a new issue to improve frequency analysis by applying time-frequency methods and merge this PR if you, the reviewers, agree!

**Review by Gui-FernandesBR — APPROVED** [review]
Extremally happy with the evolution of this PR, ready to merge.

Not an expert, but shouldn't be the case of analyzing of the follow domain ? [outOfRailTime, maxDynamicPressureTime].

Take im mind that currently the angle of attack also presents similar "arbitrary" assumptions when choosing x_lim. No problem for me now.

**Review by MateusStano — APPROVED** [review]

**Pull request merged**
